Marlin/src/module/tool_change.cpp

```diff
@@ -494,225 +477,215 @@ void tool_change(const uint8_t tmp_extruder, const float fr_mm_s/*=0.0*/, bool n
           active_extruder = tmp_extruder;
-
-        #endif // !DUAL_X_CARRIAGE
-
-        // Tell the planner the new "current position"
-        sync_plan_position();
-
-        #if ENABLED(DELTA)
-          //LOOP_XYZ(i) update_software_endstops(i); // or modify the constrain function
-          const bool safe_to_move = current_position[Z_AXIS] < delta_clip_start_height - 1;
+          update_software_endstops(X_AXIS);
+          active_extruder = !tmp_extruder;
+          const float minx = soft_endstop_min[X_AXIS], maxx = soft_endstop_max[X_AXIS];
         #else
-          constexpr bool safe_to_move = true;
+          // No software endstops? Use the configured limits
+          const float minx = tmp_extruder ? X2_MIN_POS : X1_MIN_POS,
+                      maxx = tmp_extruder ? X2_MAX_POS : X1_MAX_POS;
         #endif
 
-        // Raise, move, and lower again
-        if (safe_to_move && !no_move && IsRunning()) {
-          #if DISABLED(SWITCHING_NOZZLE)
-            // Do a small lift to avoid the workpiece in the move back (below)
-            current_position[Z_AXIS] += toolchange_settings.z_raise;
-            planner.buffer_line(current_position, planner.settings.max_feedrate_mm_s[Z_AXIS], active_extruder);
-          #endif
-          #if ENABLED(DEBUG_LEVELING_FEATURE)
-            if (DEBUGGING(LEVELING)) DEBUG_POS("Move back", destination);
-          #endif
-          #if ENABLED(DUAL_X_CARRIAGE)
-            // Dual x carriage does not properly apply these to current position due to command ordering
-            // So we apply the offsets for y and z to the destination here. X cannot have an offset in this mode
-            // as it is utilized for X2 home position.
-            destination[Y_AXIS] -= hotend_offset[Y_AXIS][active_extruder] - hotend_offset[Y_AXIS][tmp_extruder];
-            destination[Z_AXIS] -= hotend_offset[Z_AXIS][active_extruder] - hotend_offset[Z_AXIS][tmp_extruder];
+        // Don't move the new extruder out of bounds
+        if (!WITHIN(current_position[X_AXIS], minx, maxx)) no_move = true;
+
+      #endif
+
+      if (!no_move) {
+        set_destination_from_current();
+        #if DISABLED(SWITCHING_NOZZLE)
+          // Do a small lift to avoid the workpiece in the move back (below)
+          #if ENABLED(TOOLCHANGE_PARK)
+            current_position[X_AXIS] = toolchange_settings.change_point.x;
+            current_position[Y_AXIS] = toolchange_settings.change_point.y;
           #endif
-          // Move back to the original (or tweaked) position
-          do_blocking_move_to(destination[X_AXIS], destination[Y_AXIS], destination[Z_AXIS]);
-          #if ENABLED(DUAL_X_CARRIAGE)
-            active_extruder_parked = false;
+          current_position[Z_AXIS] += toolchange_settings.z_raise;
+          #if HAS_SOFTWARE_ENDSTOPS
+            NOMORE(current_position[Z_AXIS], soft_endstop_max[Z_AXIS]);
+            update_software_endstops(Z_AXIS);
```

Comment:
It makes no sense here, for example, to update software endstops if none of the relevant factors that determine software endstops has changed. And if there were such factors, why would we constrain Z before updating the software endstop for Z?